NOAA-GSL/regional_workflow
# Add workflow end to end tests for C.I. to run on AWS

## DESCRIPTION OF CHANGES: 

A Github Actions workflow file is added which will execute the specified WE2E tests defined in `tests/WE2E/machine_suites/ci.txt` on an AWS ParallelCluster configured as a Github Actions self-hosted runner. This workflow will only be initiated if the workflow run in `.github/workflows/srwa_build.yml` is successful and the `ci-aws-intel-WE` label is applied to a PR to the `rrfs_ci` branch of this repository. 

A `modulefiles/tasks/aws` directory has been added, as well as `ush/machine/aws.sh` and a couple other necessary modifications to add `aws` as a valid machine. 

The `get_single_expt_status.sh` script has been added to check each WE2E test for SUCCESS/FAILURE, polling the latest output ("Workflow status:  ") from `log.launch_FV3LAM_wflow` every 2 minutes. If a WE2E test fails or succeeds, the `get_single_expt_status.sh` will exit appropriately, clearly represented in the Github Actions UI. 

## TESTS CONDUCTED: 
The Bulid & WE2E Tests have been tested on this PR on a self-hosted runner. Results are here: 
[WE2E_tests](https://github.com/NOAA-GSL/regional_workflow/actions/runs/2701589535)
[Build_SRWA GHA Workflow](https://github.com/NOAA-GSL/regional_workflow/actions/runs/2701524335)

The proposed Continuous Integration capabilities (WE2E Tests) are tested on this PR when the `ci-aws-intel-WE` label is attached.  


## DEPENDENCIES:

- NOAA-GSL/ufs-srweather-app/pull/170

## DOCUMENTATION:

Updates to `ufs-srweather-app/docs/UsersGuide/source/ContinuousIntegration.rst` have been added. Linked PR is above.

## How to run the WE2E Tests:

After the `ci-aws-intel-build` label is attached to the PR, and the Build SRWA workflow completes, a code manager must apply the `ci-aws-intel-WE` label to run the WE2E tests. 

When the `ci-aws-intel-build` label is attached to a PR to the `rrfs_ci` branch, the build test will kick off. If the Build SRWA workflow is successful, a run_id.txt file will be saved to the ${{ github.workspace }} on the self hosted runner.  The output of the run_id.txt file provides a check to the subsequent WE2E_tests workflow to ensure the SRWA Build workflow was successful. 

The way the code is currently written for the Build SRWA workflow test, the directory in which the ufs-srweather-app is cloned and built is specified by the Build SRWA workflow's `github.run_id`.  So, the the WE2E_tests GHA workflow is coupled to the Build SRWA workflow, in the form of the `github.run_id`  in order to enter the correct directory and run the tests. 

There is no current option to have a [`workflow_run`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run) event trigger a workflow in conjunction with a label, so we are forced to attach two seperate labels for the two workflows we have (Build SRWA & WE2E_tests). 

**Iteration on how we apply labels and initiate the WE2E_tests may be necessary to achieve an ideal initiation of the WE2E tests from the Github Actions UI.** 

## CONTRIBUTORS: 
@christinaholtNOAA @venitahagerty @christopherwharrop-noaa
